app/controllers/application_controller.rb

```diff
@@ -35,8 +35,17 @@ class MissingParameterError < Exception; end
 
   def set_language
     set_gettext_locale
-    unless LANGUAGES.include? FastGettext.locale
-      params[:locale] = FastGettext.default_locale
+    requested_locale = FastGettext.locale
+    unless LANGUAGES.include? requested_locale
+      # XXX: if the browser sent de_DE, try de
+      l = requested_locale.split('_')
+      if LANGUAGES.include?l[0]
+        params[:locale] = l[0]
+      else
+        # not sure we need this. Gettext handles missing locale
+        # just fine
+        params[:locale] = FastGettext.default_locale
+      end
       set_gettext_locale
```

Comment:
I left the default out because of your comment. If you think it's not needed, be brave and remove it.

I assumed we didn't use variants with underscore at all, I missunderstood that.

But anyway, you get what I mean, don't be afraid to refactor the whole think a little but further.